### ISSUE TITLE
Feature request: Z.random

### ISSUE THREAD

**Issue by dieggsy** [ocaml/Zarith#99]
I recently added an [OCaml implementation](https://rosettacode.org/wiki/Miller%E2%80%93Rabin_primality_test#OCaml) of the [Miler-Rabin primality test](https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test) on Rosetta Code, but I find it kind of unfortunate that I have to depend on Jane Street's Bignum library for the sole purpose of generating a random bignum.

Is this something that could be added to Zarith in the future?

**Comment by xavierleroy** [comment]
The problem is to give the user control on the randomness source used.  For cryptographic applications of random prime numbers (RSA, Diffie-Hellman, etc) it is crucial to use a random source of cryptographic quality.  The `Bignum.random` function from Jane Street silently uses the Random module from the OCaml standard library, which is completely inadequate for cryptography.  

The [Cryptokit library](https://github.com/xavierleroy/cryptokit) provides cryptographically-strong randomness sources, and uses them internally to generate random Zarith bignums for RSA key generation and DH key exchange.  The corresponding functions could be exported, but I'm not sure they are as general as you'd need.

In summary: it's an API design problem.  Where do the random bits come from?  and what would be the spec for `Z.random`?

**Comment by dieggsy** [comment]
@xavierleroy wow, those are good questions and I can see why it's not entirely straightforward. For my use case I actually don't care about crypto - I'd just like to see *some* baseline random procedure for non-crypto purposes. I'd suggest that one possible route would be to emulate the API of the Random module with a disclaimer that it's not fit for cryptography, especially since there is an existing library that provides that (which could even be linked to in the docs if necessary), but I could understand objections to this as well.

**Comment by antoinemine** [comment]
Sorry for the delay, We could also wrap GMP's random functions? Not necessarily cryptographic-strength, but well-documented.

**Comment by 4y8** [comment]
> We could also wrap GMP's random functions?
I'd like to help, but I'm not sure how to represent `gmp_randstate_t`. Should I just, create a type `randstate` or do you think there is a better way of doing it. And should the random functions be in a submodule `Z.random` or just in the main module.
Thanks

**Comment by xavierleroy** [comment]
Frankly, I'm not impressed by GMP's pseudo-random generators.  We could do better just by building on OCaml's Random module.  We can also provide a function that makes it easy to use /dev/random or /dev/urandom.  I'll try to make a PR along these lines in the coming weeks.

**Comment by xavierleroy** [comment]
Should be addressed by #128.  Please re-open this feature request if something is missing.
